packages/services/src/providers/price/coingecko.provider.ts

```diff
@@ -0,0 +1,432 @@
+import { withRetry } from "../../lib/retry-helper.js";
+import {
+  BlockchainType,
+  CoinGeckoMode,
+  PriceReport,
+  PriceSource,
+  SpecificChain,
+  SpecificChainTokens,
+  TokenInfo,
+} from "../../types/index.js";
+
+/**
+ * Configuration for the CoinGecko provider
+ */
+export interface CoinGeckoProviderConfig {
+  apiKey: string;
+  mode: CoinGeckoMode;
+  specificChainTokens: SpecificChainTokens;
+}
+
+/**
+ * CoinGecko network identifiers
+ */
+type CoinGeckoNetwork =
+  | "eth"
+  | "polygon_pos"
+  | "bsc"
+  | "arbitrum"
+  | "optimism"
+  | "avax"
+  | "base"
+  | "linea"
+  | "zksync"
+  | "scroll"
+  | "mantle"
+  | "solana";
+
+/**
+ * CoinGecko network mapping from `SpecificChain` to CoinGecko network identifier
+ */
+const COINGECKO_NETWORKS = {
+  eth: "eth",
+  polygon: "polygon_pos",
+  bsc: "bsc",
+  arbitrum: "arbitrum",
+  optimism: "optimism",
+  avalanche: "avax",
+  base: "base",
+  linea: "linea",
+  zksync: "zksync",
+  scroll: "scroll",
+  mantle: "mantle",
+  svm: "solana",
+} as const satisfies Record<SpecificChain, CoinGeckoNetwork>;
+
+/**
+ * Subset of the CoinGecko onchain response used in the price provider
+ */
+const OnchainResponseSchema = z.object({
+  data: z.object({
+    attributes: z.object({
+      symbol: z.string(),
+      price_usd: z.string(),
+      volume_usd: z.object({
+        h24: z.string(),
+      }),
+      total_reserve_in_usd: z.string(),
+      fdv_usd: z.string(),
+    }),
+  }),
+  included: z.array(
+    z.object({
+      attributes: z.object({
+        pool_created_at: z.string(),
+      }),
+    }),
+  ),
+});
+
+/**
+ * CoinGecko price provider implementation
+ * Provides cryptocurrency price data using the CoinGecko API
+ */
+export class CoinGeckoProvider implements PriceSource {
+  private readonly client: Coingecko;
+  private readonly MAX_RETRIES = 3; // Overrides default of 2
+  private readonly MAX_TIMEOUT = 30_000; // Overrides default of 60 seconds
+  private readonly BATCH_CONCURRENCY_LIMIT = 30; // Process batch calls in max 30 tokens at a time
+  private specificChainTokens: SpecificChainTokens;
+  private logger: Logger;
+  private coingeckoLogger: Logger; // Native logger for the CoinGecko SDK (set log level via `COINGECKO_LOG`)
+
+  /**
+   * Creates a new CoinGecko provider instance
+   * Initializes the CoinGecko client with appropriate configuration based on environment
+   * @param config - Provider configuration including API key, mode, and chain tokens
+   * @param logger - Logger instance for error and debug logging
+   */
+  constructor(config: CoinGeckoProviderConfig, logger: Logger) {
+    this.specificChainTokens = config.specificChainTokens;
+    this.logger = logger;
+    this.coingeckoLogger = logger.child({ context: "CoinGeckoSDK" });
+    const { apiKey, mode } = config;
+    const opts: ClientOptions = {
+      environment: mode,
+      maxRetries: this.MAX_RETRIES,
+      timeout: this.MAX_TIMEOUT,
+      logger: this.coingeckoLogger,
+    };
```

Comment:
we use the coingecko SDK's native retry and timeout logic. 

however, this doesnt account for 429s. you'll see below that i wrap method calls in `withRetry` to account for this. alternatively, we could disable the coingecko SDK's logic by setting `maxRetries` and `timeout`, and rely purely on our helper method.